README.md
# builder-webpack


[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/iv-web/feflow/blob/master/LICENSE)
[![](https://img.shields.io/travis/feflow/builder-webpack3.svg)](https://travis-ci.org/feflow/builder-webpack3)
[![npm package](https://img.shields.io/npm/v/builder-webpack3.svg?style=flat-square)](https://www.npmjs.org/package/builder-webpack3)
[![NPM downloads](http://img.shields.io/npm/dt/builder-webpack3.svg?style=flat-square)](https://npmjs.org/package/builder-webpack3)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/cpselvis/builder-webpack3/pulls)

Webpack 构建器, 适用于NOW直播业务和活动类型的项目构建

## 特性

- 对H5开发友好，默认集成 Rem 方案，解决适配问题
- 支持 SCSS 和 JS 模块 import 时的绝对路径写法
- 支持多页面打包的开发方式
- 支持 FIS3 的 html 文件 inline 语法糖
- 支持雪碧图(图片后缀加__sprite)
- 使用Happypack 多实例构建，飞一般的构建速度

## 安装

确保feflow的版本在 v0.12.0 以上, 可以通过如下命令安装最新feflow版本
```
$ npm install feflow-cli -g
```

## 快速使用

### 添加feflow.json配置文件

在项目根目录添加 `feflow.json` 配置文件

``` sh
{
    "builderType": "builder-webpack3",
    "builderOptions": {
        "outDir": "dist",                                    // 资源输出路径，默认 public
        "product": "now",                                    // 产品，此处可以是 now 或者 shangfen
        "domain": "now.qq.com",                              // 域名，离线包的域名需要使用
        "cdn": "11.url.cn",                                  // 资源发布到的cdn名称
        "moduleName": "mobile",                              // 部署的模块
        "bizName": "category",                               // 业务名称
        "minifyHTML": true,                                  // 是否压缩 html
        "minifyCSS": true,                                   // 是否压缩 js
        "minifyJS": true,                                    // 是否压缩 css
        "inlineCSS": true,                                   // 生成的 css 是否内联到首屏
        "usePx2rem": true,                                   // 是否使用 Rem
        "useReact": true,                                    // 是否是 React，如果为false，则不会在 html 中引用 React 框架 
        "remUnit": 37.5,                                     // Rem 单位，对于 375 视觉稿，此处填写 37.5，750视觉稿需要改成 75 
        "remPrecision": 8,                                   // Rem 的精度，即 px 转换成了 rem 后的小数点后位数
        "inject": true,                                      // 打包生成的 js 文件是否自动注入到 html 文件 body 之后
        "port": 8001,                                        // 本地开发的 webpack 构建服务进程端口号
        "externals": [                                       // 基础框架不打入到 bundle 里面
            {
                "module": "react",
                "entry": "//11.url.cn/now/lib/16.2.0/react.min.js?_bid=3123",
                "global": "React"
            },
            {
                "module": "react-dom",
                "entry": "//11.url.cn/now/lib/16.2.0/react-dom.min.js?_bid=3123",
                "global": "ReactDOM"
            }
        ]
    }
}
```

### 命令

```sh
$ feflow dev      # 本地开发时的命令
$ feflow build    # 发布时的打包命令, 打出的包在工程的public目录, 包含 cdn, webserver 和 offline 三个文件夹
```

## 文档

### 内联

同时支持Fis3项目的inline语法糖写法和ejs的写法

- 内联 html:

``` sh
<!--inline[/assets/inline/meta.html]-->
```

- 内联 javascript

``` sh
<script src="@tencent/report-whitelist?__inline"></script>
```

备注：如果希望内联某个 JS 文件，需要使用相对路径的写法。

### 代理设置

- 执行 `feflow dev` 命令后会在本地的 8001 端口开启一个 WDS 服务，所有的静态资源(html, css, js, img) 都会在内存里面。可以通过 http://127.0.0.1:8001/webpack-dev-server  查看

![](https://qpic.url.cn/feeds_pic/ajNVdqHZLLDpvNiayyEbzqB9V61CRiallnRdEKFaViaxw7pibicBKgEI8vw/)

- Fiddler配置把之前的本地绝对路径改成 本地server 路径即可：

![](https://qpic.url.cn/feeds_pic/Q3auHgzwzM72dIPZyXSdy8srwzIOTovf0VSaNlBzE98ueBiaibIVSHkA/)

### 热更新支持

- 如果要支持热更新，需要再增加一条代理`_webpack_hmr`的配置，如：

`/^https?://now\.qq\.com/(__webpack_hmr)$/ http://127.0.0.1:8001/$1`

- 在项目中，用`react-hot-loader`将`pageComponent`变为可接受热更新的组件

```js
import { hot } from 'react-hot-loader'
class pageComponent extends Component {
    ...
}
export default hot(module)(pageComponent)
```

### 测试

1. `git clone`这个用于烟雾测试的[模板项目](https://github.com/feflow/generator-smoking-test)
2. 配置`.travis.yml`，可以参考模板项目[README](https://github.com/feflow/generator-smoking-test)
3. 在[Travis-ci](https://travis-ci.org/feflow/builder-webpack3)中打开此项目的自动构建

## 版本日志

[版本日志](CHANGELOG.md)

## 许可证

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
<a name="0.4.19"></a>
## [0.4.19](https://github.com/iv-web/builder-webpack/compare/v0.4.16...v0.4.19) (2019-04-24)


### Bug Fixes

* remove pic compress ([d8c401c](https://github.com/iv-web/builder-webpack/commit/d8c401c))
* 修复上下文bug ([8a5d621](https://github.com/iv-web/builder-webpack/commit/8a5d621))
* 锁住html-webpack-include-assets-plugin的版本号 ([d0e6237](https://github.com/iv-web/builder-webpack/commit/d0e6237))



<a name="0.4.15"></a>
## [0.4.15](https://github.com/iv-web/builder-webpack/compare/v0.4.13...v0.4.15) (2019-03-06)


### Bug Fixes

* 修复静态资源离线包的bug ([097d05f](https://github.com/iv-web/builder-webpack/commit/097d05f))


### Features

* 错误构建阶段抛出，避免发布到线上 ([89e4c19](https://github.com/iv-web/builder-webpack/commit/89e4c19))



<a name="0.4.12"></a>
## [0.4.12](https://github.com/iv-web/builder-webpack/compare/v0.4.11...v0.4.12) (2018-12-31)


### Bug Fixes

* fix dev mode loader path ([84d635e](https://github.com/iv-web/builder-webpack/commit/84d635e))
* 修复开发环境下的图片重复问题 ([c107356](https://github.com/iv-web/builder-webpack/commit/c107356))



<a name="0.4.10"></a>
## [0.4.10](https://github.com/iv-web/builder-webpack/compare/v0.4.9...v0.4.10) (2018-12-15)


### Features

* support global-style loader path ([f5a3e7c](https://github.com/iv-web/builder-webpack/commit/f5a3e7c))



<a name="0.4.9"></a>
## [0.4.9](https://github.com/iv-web/builder-webpack/compare/v0.4.7...v0.4.9) (2018-11-22)


### Bug Fixes

* 支持 ios8 机型的 autoprefix ([610fb4a](https://github.com/iv-web/builder-webpack/commit/610fb4a))


### Features

* 支持css热更新 ([dcd7fa6](https://github.com/iv-web/builder-webpack/commit/dcd7fa6))



<a name="0.4.7"></a>
## [0.4.7](https://github.com/iv-web/builder-webpack/compare/v0.4.6...v0.4.7) (2018-09-20)


### Features

* 支持设置outDir ([6769eb8](https://github.com/iv-web/builder-webpack/commit/6769eb8))



<a name="0.4.6"></a>
## [0.4.6](https://github.com/iv-web/builder-webpack/compare/v0.4.5...v0.4.6) (2018-09-11)


### Bug Fixes

* 删除掉 Webview 预加载的插件 ([761759f](https://github.com/iv-web/builder-webpack/commit/761759f))


### Features

* 增加测试用模板，以及相关测试代码 ([31af928](https://github.com/iv-web/builder-webpack/commit/31af928))
* 更新package.json，更新test.js，删除public ([63c893e](https://github.com/iv-web/builder-webpack/commit/63c893e))



<a name="0.4.5"></a>
## [0.4.5](https://github.com/iv-web/builder-webpack/compare/v0.4.4...v0.4.5) (2018-09-03)


### Bug Fixes

* node_modules里面的内容不编译. ([b50f7b5](https://github.com/iv-web/builder-webpack/commit/b50f7b5))



<a name="0.4.4"></a>
## [0.4.4](https://github.com/iv-web/builder-webpack/compare/v0.4.3-alpha.0...v0.4.4) (2018-08-24)


### Features

* 增加less支持. ([23f633a](https://github.com/iv-web/builder-webpack/commit/23f633a))
* 增加tree shaking特性. ([111579f](https://github.com/iv-web/builder-webpack/commit/111579f))



<a name="0.4.3"></a>
## [0.4.3](https://github.com/iv-web/builder-webpack/compare/v0.4.3-alpha.0...v0.4.3) (2018-08-23)


### Features

* 增加less支持. ([23f633a](https://github.com/iv-web/builder-webpack/commit/23f633a))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/iv-web/builder-webpack/compare/v0.3.36...v0.4.0) (2018-08-21)


### Bug Fixes

* 修改注释，去除不必要依赖 ([d5d8c9f](https://github.com/iv-web/builder-webpack/commit/d5d8c9f))
* 开放builder-webpack-core版本 ([08457bf](https://github.com/iv-web/builder-webpack/commit/08457bf))

### Features

* 兼容原来的代理配置，去除不必要的依赖库 ([c0b752f](https://github.com/iv-web/builder-webpack/commit/c0b752f))
* 支持热更新，开发环境使用publicPath方便配代理 ([873a038](https://github.com/iv-web/builder-webpack/commit/873a038))


<a name="0.3.34"></a>
## [0.3.34](https://github.com/iv-web/builder-webpack/compare/v0.3.33...v0.3.34) (2018-07-04)


### Bug Fixes

* feflow.js导致的路径问题修复. ([042ff66](https://github.com/iv-web/builder-webpack/commit/042ff66))



<a name="0.3.33"></a>
## [0.3.33](https://github.com/iv-web/builder-webpack/compare/v0.3.32...v0.3.33) (2018-07-04)


### Bug Fixes

* 雪碧图修复. ([e4d7537](https://github.com/iv-web/builder-webpack/commit/e4d7537))



<a name="0.3.32"></a>
## [0.3.32](https://github.com/iv-web/builder-webpack/compare/v0.3.31...v0.3.32) (2018-07-04)


### Features

* 支持feflow.js配置文件. ([9e94898](https://github.com/iv-web/builder-webpack/commit/9e94898))



<a name="0.3.21"></a>
## [0.3.21](https://github.com/iv-web/builder-webpack/compare/v0.3.18...v0.3.21) (2018-06-20)


### Bug Fixes

* 雪碧图bug修复. ([ed78054](https://github.com/iv-web/builder-webpack/commit/ed78054))


### Features

* 增加对TS语法的支持。 ([4e0dae7](https://github.com/iv-web/builder-webpack/commit/4e0dae7))



<a name="0.3.18"></a>
## [0.3.18](https://github.com/iv-web/builder-webpack/compare/v0.3.17...v0.3.18) (2018-05-17)


### Features

* code split, 抽离出公共js文件 ([9b577af](https://github.com/iv-web/builder-webpack/commit/9b577af))



<a name="0.3.16"></a>
## [0.3.16](https://github.com/iv-web/builder-webpack/compare/v0.3.15...v0.3.16) (2018-05-10)


### Features

* 支持官网场景. ([dd641fc](https://github.com/iv-web/builder-webpack/commit/dd641fc))



<a name="0.3.15"></a>
## [0.3.15](https://github.com/iv-web/builder-webpack/compare/v0.3.14...v0.3.15) (2018-05-09)


### Features

* 支持不传入bizName, 使用于 now.qq.com/index.html 官网的场景. ([93f932d](https://github.com/iv-web/builder-webpack/commit/93f932d))



<a name="0.3.14"></a>
## [0.3.14](https://github.com/iv-web/builder-webpack/compare/v0.3.13...v0.3.14) (2018-04-28)


### Features

* 捷豹部署平台的Loader路径. ([5b99f63](https://github.com/iv-web/builder-webpack/commit/5b99f63))



<a name="0.3.13"></a>
## [0.3.13](https://github.com/iv-web/builder-webpack/compare/v0.3.12...v0.3.13) (2018-04-28)


### Bug Fixes

* 增加对现有版本的兼容. ([2917325](https://github.com/iv-web/builder-webpack/commit/2917325))



<a name="0.3.12"></a>
## [0.3.12](https://github.com/iv-web/builder-webpack/compare/v0.3.10...v0.3.12) (2018-04-27)


### Bug Fixes

* externals去掉http schema的头部. ([cc8d351](https://github.com/iv-web/builder-webpack/commit/cc8d351))

test/template-project/test.js
// const assert = require('assert');
const assert = require("assert");
const webpack = require("webpack");
const glob = require("glob-all");
const rimraf = require('rimraf');

const testFileExists = function() {
    describe("smoking test template", function() {
      
      it("should generate webserver including html files", function(done) {
        glob(
          ["./public/webpack/index.html"],
          err => {
            if (!err) {
              done();
            }
          }
        );
      });
      it("should generate category dirctory including js & css files", function(done) {
        glob(
          ["./public/cdn/category/index_*.js", "./public/cdn/category/idnex_*.css"],
          err => {
            if (!err) {
              done();
            }
          }
        );
      });
      it("should generate zip for offline bundle", function(done) {
          glob('./public/offline/offline.zip', err => {
              if (!err) {
                  done()
              }
          })
      })
    });
}


describe("building based on test template", function() {
    it('generate a project with webpack in production mode', function(done) {
        process.chdir(__dirname); // 切换到当前template-project，否则找不到feflow.json
        const builder = require("../../lib/index");
        rimraf('./public', () => {
            webpack(builder.prodConfig, (err, stats) => {
                console.log(stats.toString({
                    chunks: false,
                    children: false
                }))
                done()
                testFileExists()
            })
        })
    });
})



test/template-project/index.js
const builder = require('../../lib/index')

builder('build')

test/template-project/webpack.dev.js
module.exports = require('../../lib/index').devConfig;

lib/util.js
'use strict';

const fs = require('fs');
const path = require('path');

/**
 * 深拷贝, Object.assign()只有第一层是深拷贝, 第二层之后仍然是 浅拷贝
 * @param source
 * @returns {{}}
 */
const deepCopy = (source) => {
    const ret = {};

    for (var k in source) {
        ret[k] = typeof source[k] ==='object' ? deepCopy(source[k]) : source[k]
    }
    return ret;
};

/**
 * 列出某个目录下的子目录, DFS算法
 * @param root         目录路径
 * @param level        列出的子目录层级
 * @param directories  默认为[]
 * @returns {*|Array}
 */
const listDir =(root, level, directories) => {
    directories = directories || [];

    if (!fs.existsSync(root)) {
        return directories;
    }

    if (fs.statSync(root).isDirectory() && level > 0) {
        fs.readdirSync(root)
            .forEach((name) => {
                const dirPath = path.join(root, name);
                if (fs.statSync(dirPath).isDirectory()) {
                    directories.push({
                        name,
                        dirPath
                    });
                    listDir(dirPath, level - 1, directories);
                }
            })
    }

    return directories;
}

exports.deepCopy = deepCopy;
exports.listDir = listDir;


lib/config.js
'use strict';

const fs = require('fs');
const path = require('path');

class Config {
    /**
     * @function getPath
     * @desc     Find feflow.json file
     */
    static getPath() {
        let currDir = process.cwd();

        while (!fs.existsSync(path.join(currDir, 'feflow.json'))) {
            currDir = path.join(currDir, '../');

            // unix跟目录为/， win32系统根目录为 C:\\格式的
            if (currDir === '/' || /^[a-zA-Z]:\\$/.test(currDir)) {
                console.error('未找到 feflow.json');
                process.exit(1);
            }
        }

        return currDir;
    }

    /**
     * @function getBuildConfig
     * @desc     Find builder type in feflow.json
     */
    static getBuildConfig() {
        const configFile = path.join(Config.getPath(), './feflow.json');

        if (!fs.existsSync(configFile)) {
            throw new Error('未找到 feflow.json');
        } else {
            const fileContent = fs.readFileSync(configFile, 'utf-8');

            let feflowCfg;

            try {
                feflowCfg = JSON.parse(fileContent);
            } catch (ex) {
                throw new Error('请确保feflow.json配置是一个Object类型，并且含有builderType字段');
            }

            return feflowCfg.builderOptions;
        }
    }
}

module.exports = Config;


lib/server.js
"use strict";

const express = require("express");
const webpack = require("webpack");
const webpackDevMiddleware = require("webpack-dev-middleware");
const webpackHotMiddleware = require("webpack-hot-middleware");
const Config = require("./config");

const app = express();

module.exports = devConfig => {
  // 添加webpack hmr入口，这要求项目中也要安装webpack-hot-middleware，否则webpack找不到该模块
  // TODO: 这里其实可以添加resolve解决
  for (var key in devConfig.entry) {
    devConfig.entry[key] = [
      `webpack-hot-middleware/client?dynamicPublicPath=true&path=__webpack_hmr`,
      devConfig.entry[key]
    ];
    // 如果是react-hot-loader 3.0，这一行可以注释
    // devConfig.entry[key].unshift("react-hot-loader/patch");
  }
  const compiler = webpack(devConfig);

  // 配置devServer
  app.use(
    webpackDevMiddleware(compiler, {
      // publicPath: devConfig.output.publicPath, // 默认就会使用devConfig的publicPath
      hot: true,
      color: true,
      stats: "errors-only" // 为了减少webpack不必要的输出，将stats设为errors-only
    })
  );

  // 加入热更新中间件
  app.use(webpackHotMiddleware(compiler));


  // Serve the files on port.
  app.listen(devConfig.devServer.port, function(res, err) {
    if (err) {
      console.error(err);
    } else {
      console.log(
        `Webpack server listening on port ${devConfig.devServer.port}\n`
      );
    }
  });
};


lib/builder.js
'use strict';

/**
 * Copyright (c) 2018 Tencent Inc.
 *
 * Webpack构建器，适用于NOW直播IVWEB团队工程项目.
 *
 * cpselvis <cpselvis@gmal.com>
 */
const path = require('path');
const glob = require("glob");
const webpack = require('webpack');
const CleanWebpackPlugin = require('clean-webpack-plugin');
const HtmlWebpackInlineSourcePlugin = require('html-webpack-inline-source-plugin');
const StringReplaceWebpackPlugin = require('string-replace-webpack-plugin');
const BuilderCore = require('builder-webpack-core');
const { deepCopy } = require('./util');
const Config = require('./config');

// 当前运行的时候的根目录
let projectRoot = Config.getPath('feflow.json');

if (!projectRoot) {
    projectRoot = Config.getPath('feflow.js');
}

// 最基础的配置
const baseConfig = {
    entry: glob.sync(path.join(projectRoot, './src/pages/*')),
    module: {
        rules: []
    },
    plugins: [],
    resolve: {
        alias: glob.sync(path.join(projectRoot, './src/*/')) // 支持Webpack import绝对路径的写法
    },
    resolveLoader: {}
};

class Builder extends BuilderCore {

    /**
     * @function createDevConfig
     * @desc     创建用于开发过程中的webpack打包配置
     *
     * @param {Object}  options                         参数
     * @param {Boolean} options.usePx2rem               是否启用px2rem
     * @param {Number}  options.remUnit                 px2rem的单位, 默认75
     * @param {Number}  options.remPrecision            px2rem的精度，默认8
     * @param {Number}  options.port                    webpack打包的端口号，默认8001
     * @param {String}  options.inject                  是否注入chunks
     *
     * @example
     */
    static createDevConfig(options) {
        const devConfig = deepCopy(baseConfig);

        // 设置打包规则
        const devRules = [];
        // 设置HTML解析规则
        devRules.push(this.setHtmlRule());
        // 设置图片解析规则
        devRules.push(this.setImgRule(true));
        // 设置CSS解析规则
        devRules.push(this.setCssRule());
        // 设置Less解析规则，开发环境不压缩CSS
        devRules.push(this.setLessRule(false, options.usePx2rem, options.remUnit, options.remPrecision));
        // 设置SCSS解析规则，开发环境不压缩CSS
        devRules.push(this.setScssRule(false, options.usePx2rem, options.remUnit, options.remPrecision));
        // 设置JS解析规则
        devRules.push(this.setJsRule());
        // 设置TS解析规则
        devRules.push(this.setTsRule());
        // 设置字体解析规则
        devRules.push(this.setFontRule());

        // 设置打包插件
        let devPlugins = [];
        devPlugins.push(new StringReplaceWebpackPlugin());
        // 设置提取CSS为一个单独的文件的插件
        devPlugins.push(this.setExtractTextPlugin(false, ''));

        if (options.useReact !== false) {
            // React, react-dom 通过cdn引入
            devPlugins.push(this.setExternalPlugin(options.externals));
        }
        // 设置NODE_ENV 为 development
        devPlugins.push(this.setDefinePlugin('development'));
        // 多实例构建
        devPlugins.push(this.setHappyPack(options.jsLoader, options.useTreeShaking));
        // 增加热更新组件
        devPlugins.push(new webpack.HotModuleReplacementPlugin());
        // 抽离公共js
        // devPlugins.push(this.setCommonsChunkPlugin());
        // 多页面打包

        const {newEntry, htmlWebpackPlugins} = this.setMultiplePage(devConfig.entry, false, options.inject, false, '', '');
        devPlugins = devPlugins.concat(htmlWebpackPlugins);

        devConfig.entry = newEntry;
        // 开发阶段增加sourcemap.
        devConfig.devtool = 'inline-source-map';
        // 这里还是依然按照原来的配置，将静态资源用根目录伺服
        devConfig.output = this.setOutput(false, '', '/', options.outDir);
        devConfig.module.rules = devRules;
        devConfig.plugins = devPlugins;
        devConfig.devServer = this.setDevServer(options.port || 8001);
        devConfig.resolve.alias = this.setAlias(options.alias);
        devConfig.resolve.extensions = ['.js', '.jsx', '.ts', '.tsx', '.json'];
        // 设置 loader 的npm包查找的相对路径，此处设置在全局的 .feflow 目录下
        devConfig.resolveLoader = this.setResolveLoaderPath('builder-webpack3', options.runtime);

        return devConfig;
    }

    /**
     * @function createProdConfig
     * @desc     创建用于生产环境中的webpack打包配置
     *
     * @param {Object}  options                         参数
     * @param {Boolean} options.minifyHTML              是否压缩HTML
     * @param {Boolean} options.minifyCSS               是否压缩CSS
     * @param {Boolean} options.minifyJS                是否压缩JS
     * @param {Boolean} options.usePx2rem               是否启用px2rem
     * @param {Number}  options.remUnit                 px2rem的单位, 默认75
     * @param {Number}  options.remPrecision            px2rem的精度，默认8
     * @param {String}  options.moduleName              模块名称 (如果不传，则放在根目录)
     * @param {String}  options.bizName                 业务名称
     * @param {String}  options.inject                  是否注入chunks
     * @example
     */
    static createProdConfig(options) {
        const prodConfig = deepCopy(baseConfig);

        const bizName = options.bizName;
        const moduleName = options.moduleName;
        // 业务域名
        const domain = options.domain || 'now.qq.com';
        const cdn = options.cdn || '11.url.cn';
        const product = options.product || 'now';
        // Html 路径前缀, 打包时的目录
        const htmlPrefix = moduleName ? `webserver/${bizName}` : 'webserver';
        // Css, Js, Img等静态资源路径前缀, 打包时的目录
        const assetsPrefix = moduleName ? `cdn/${bizName}` : 'cdn';
        const cdnUrl = moduleName ? `//${cdn}/${product}/${moduleName}/${bizName}` : `//${cdn}/${product}/${bizName}`;
        const serverUrl = moduleName ? `//${domain}/${moduleName}/${bizName}` : `//${domain}/${bizName}`;

        const regex = new RegExp(assetsPrefix + '/', 'g');

        // 设置打包规则
        const prodRules = [];
        // 设置HTML解析规则
        prodRules.push(this.setHtmlRule());
        // 设置图片解析规则, 图片需要hash
        prodRules.push(this.setImgRule(true, assetsPrefix));
        // 设置CSS解析规则
        prodRules.push(this.setCssRule());
        // 设置Less解析规则，生产环境默认压缩CSS
        prodRules.push(this.setLessRule(false, options.usePx2rem, options.remUnit, options.remPrecision));
        // 设置SCSS解析规则，生产环境默认压缩CSS
        prodRules.push(this.setScssRule(options.minifyCSS || true, options.usePx2rem, options.remUnit, options.remPrecision));
        // 设置JS解析规则
        prodRules.push(this.setJsRule());
        // 设置TS解析规则
        prodRules.push(this.setTsRule());
        // 设置字体解析规则
        prodRules.push(this.setFontRule());

        // 设置打包插件
        let prodPlugins = [];
        // 清空Public目录插件, https://github.com/johnagan/clean-webpack-plugin/issues/17
        prodPlugins.push(new CleanWebpackPlugin([options.outDir], {
            root: projectRoot,
            verbose: true,
            dry: false
        }));
        prodPlugins.push(new StringReplaceWebpackPlugin());
        // 设置提取CSS为一个单独的文件的插件
        prodPlugins.push(this.setExtractTextPlugin(true, assetsPrefix));
        if (options.minifyJS) {
            // 压缩JS
            prodPlugins.push(new webpack.optimize.UglifyJsPlugin());
        }
        if (options.useReact !== false) {
            // React, react-dom 通过cdn引入
            prodPlugins.push(this.setExternalPlugin(options.externals));
        }
        // Inline 生成出来的css
        prodPlugins.push(new HtmlWebpackInlineSourcePlugin());
        // 设置NODE_ENV 为 production
        prodPlugins.push(this.setDefinePlugin('production'));
        // 多实例构建
        prodPlugins.push(this.setHappyPack(options.jsLoader, options.useTreeShaking));
        // 抽离公共js
        //prodPlugins.push(this.setCommonsChunkPlugin());
        // 支持Fis3的 inline 语法糖 多页面打包, 默认压缩html
        const {newEntry, htmlWebpackPlugins} = this.setMultiplePage(prodConfig.entry, options.minifyHTML, options.inject, options.inlineCSS, assetsPrefix, htmlPrefix);

        prodPlugins = prodPlugins.concat(htmlWebpackPlugins);
        // 修复html引用css和js的路径问题
        prodPlugins.push(this.setReplaceHtmlPlugin(regex, ''));
        // 修复js和css中引用的图片路径问题
        prodPlugins.push(this.setReplaceBundlePlugin(regex, ''));
        if (options.useSri !== false) {
            // 给生成出来的js bundle增加跨域头(cross-origin)，便于错误日志记录
            prodPlugins.push(this.setSriPlugin());
        }

        prodPlugins.push(this.setOffline(assetsPrefix, htmlPrefix, cdnUrl, serverUrl, domain, cdn, product, options.outDir));

        prodPlugins.push(this.setDoneErrorPlugin());

        prodConfig.entry = newEntry;
        prodConfig.output = this.setOutput(true, assetsPrefix, cdnUrl + '/', options.outDir);
        prodConfig.module.rules = prodRules;
        prodConfig.plugins = prodPlugins;
        prodConfig.resolve.alias = this.setAlias(options.alias);
        prodConfig.resolve.extensions = ['.js', '.jsx', '.ts', '.tsx', '.json'];
        // 设置 loader 的npm包查找的相对路径，此处设置在全局的 .feflow 目录下
        prodConfig.resolveLoader = this.setResolveLoaderPath('builder-webpack3', options.runtime);

        return prodConfig;
    }
}

module.exports = Builder;
